### PR TITLE
Add missing foreign keys to fix #545

### LINKED ISSUE #545: Egg variable name needs to be unique per egg
Currently you can create multiple variables with the same name. This should not be possible.
Same applies to the env variable name.

### PR DESCRIPTION
**Pull request by RMartinOscar**
- [x] Make name & env_variables unique for each egg

> Throws a 500 if you try to add a duplicate

### CODE REVIEW

File: database/migrations/2024_08_14_182737_fix_egg_variable_foreign_key.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+return new class extends Migration
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        // Only needed for sqlite
+        if (Schema::getConnection()->getDriverName() !== 'sqlite') {
```
Comment (Boy132) [inline database/migrations/2024_08_14_182737_fix_egg_variable_foreign_key.php:15]: Fix needs to apply to all drivers, not just sqlite

File: database/migrations/2024_08_14_182737_fix_egg_variable_foreign_key.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                "rules" text not null, 
+                "created_at" datetime, 
+                "updated_at" datetime, 
+                "sort" integer, 
+                foreign key("egg_id") references "eggs"("id") on delete cascade)');
+            DB::statement('INSERT INTO egg_variables SELECT * FROM _egg_variables_old');
+            DB::statement('DROP TABLE _egg_variables_old');
+            DB::statement('CREATE UNIQUE INDEX "egg_variables_env_variable_unique" on "egg_variables" ("egg_id", "env_variable")');
+            DB::statement('CREATE UNIQUE INDEX "egg_variables_name_unique" on "egg_variables" ("egg_id", "name")');
+        });
```
Comment (Boy132) [inline database/migrations/2024_08_14_182737_fix_egg_variable_foreign_key.php:44]: You don't need to drop and recreate the table to create an index. That is only needed when changing foreign keys on sqlite.

Create a new migration with something like this:
```php
Schema::table('egg_variables', function (Blueprint $table) {
    $table->unique(['egg_id', 'env_variable']);
    $table->unique(['egg_id', 'name']);
});
```

File: database/migrations/2024_08_14_182737_fix_egg_variable_foreign_key.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        DB::statement('PRAGMA foreign_keys = ON');
+        DB::statement('PRAGMA legacy_alter_table = OFF');
+    }
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down(): void
+    {
+        // Reverse not needed
```
Comment (Boy132) [inline database/migrations/2024_08_14_182737_fix_egg_variable_foreign_key.php:55]:
```php
Schema::table('egg_variables', function (Blueprint $table) {
    $table->dropUnique(['egg_id', 'env_variable']);
    $table->dropUnique(['egg_id', 'name']);
});
```

**Review by Boy132 — CHANGES_REQUESTED** [review]
Besides the backend changes we should also handle the frontend properly.

**Comment by RMartinOscar** [comment]
@Boy132 sqlite tests are still failling

**Comment by Boy132** [comment]
> @Boy132 sqlite tests are still failling

I found the reason: the `sort` for the bungeecord variables is `null`. For all other eggs the `sort` is filled.

**Comment by notAreYouScared** [comment]
@RMartinOscar Can you merge main, and see if the tests pass?

**Comment by lancepioch** [comment]
We should programmatically enforce this instead, I think sqlite has an issue with composite unique keys.

**Pull request closed without merging**